Ungraceful continuous testing failure in dev mode due to (unhandled?) Jackson serialization error when communicating with dev ui

### Describe the bug

When running tests in development mode (`quarkus dev` or `mvn quarkus:dev`), if an exception is thrown as part of a test that cannot be serialized to json for the dev UI, then continuous testing fails with a jackson processing exception and you have to restart the quarkus dev process.

If you run your tests with `quarkus:test` then everything is fine, as there is no attempt to parse the test results as json.

I think that any sort of error / exception that occurs in the dev-ui/as part of the dev-ui communication should not interfere with continuous testing at all. These errors (IMO) should be stopped before they bleed into the continuous processing experience, as when they do, this interrupts development flow using `quarkus dev`.

### Expected behavior

An error occurring in the communication process for the dev-ui should not cause the continuous testing process to fail, requiring a restart.

### Actual behavior

An error occurring in the communication process for the dev-ui causes the continuous testing process to fail, requiring a restart.

### How to Reproduce?

1. Download the [reproducer](https://github.com/quarkusio/quarkus/files/12484269/quarkus-dev-ui-jackson-exception-reproducer.zip)
2. Run `mvn quarkus:test`
3. See a nicely reported test failure like so:
```bash
2023-08-31 10:45:24,901 ERROR [io.qua.test] (Test runner thread) ==================== TEST REPORT #1 ==================== [Error Occurred After Shutdown]
2023-08-31 10:45:24,901 ERROR [io.qua.test] (Test runner thread) Test AwesomeServiceTest#watch it burn() failed
 [Error Occurred After Shutdown]: org.jboss.resteasy.reactive.ClientWebApplicationException: Received: 'Not Found, status code 404' when invoking: Rest Client method: 'de.awesomecompany.awesomeproduct.awesomeness.AwesomeRestClient#activateMaximumCoolness'
        at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext.unwrapException(RestClientRequestContext.java:185)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.handleException(AbstractResteasyReactiveContext.java:322)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:173)
        at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext$1.lambda$execute$0(RestClientRequestContext.java:295)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:246)
        at io.vertx.core.impl.EventLoopContext.lambda$runOnContext$0(EventLoopContext.java:43)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: jakarta.ws.rs.WebApplicationException: Not Found, status code 404
        at io.quarkus.rest.client.reactive.runtime.DefaultMicroprofileRestClientExceptionMapper.toThrowable(DefaultMicroprofileRestClientExceptionMapper.java:18)
        at io.quarkus.rest.client.reactive.runtime.MicroProfileRestClientResponseFilter.filter(MicroProfileRestClientResponseFilter.java:52)
        at org.jboss.resteasy.reactive.client.handlers.ClientResponseFilterRestHandler.handle(ClientResponseFilterRestHandler.java:21)
        at org.jboss.resteasy.reactive.client.handlers.ClientResponseFilterRestHandler.handle(ClientResponseFilterRestHandler.java:10)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.invokeHandler(AbstractResteasyReactiveContext.java:229)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
        ... 12 more
 ```
4. Stop `quarkus:test`
5. Start `quarkus:dev`
6. Press 'r' to start continuous testing process
7. See a jackson serialization error like so
```bash
2023-08-31 10:49:15,226 ERROR [io.qua.test] (Test runner thread) Internal error running tests: java.lang.RuntimeException: java.lang.RuntimeException: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class java.io.ByteArrayInputStream and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS) (through reference chain: io.quarkus.deployment.dev.testing.TestRunResults["results"]->java.util.Collections$UnmodifiableMap["AwesomeServiceTest"]->io.quarkus.deployment.dev.testing.TestClassResult["failing"]->java.util.ArrayList[0]->io.quarkus.deployment.dev.testing.TestResult["problems"]->java.util.Collections$UnmodifiableRandomAccessList[0]->org.jboss.resteasy.reactive.ClientWebApplicationException["cause"]->jakarta.ws.rs.WebApplicationException["response"]->org.jboss.resteasy.reactive.client.impl.ClientResponseImpl["entityStream"])
        at io.quarkus.devui.runtime.continuoustesting.ContinuousTestingJsonRPCService.invokeAction(ContinuousTestingJsonRPCService.java:96)
        at io.quarkus.devui.runtime.continuoustesting.ContinuousTestingJsonRPCService.getResults(ContinuousTestingJsonRPCService.java:89)
        at io.quarkus.devui.runtime.continuoustesting.ContinuousTestingJsonRPCService.accept(ContinuousTestingJsonRPCService.java:39)
        at io.quarkus.devui.runtime.continuoustesting.ContinuousTestingJsonRPCService.accept(ContinuousTestingJsonRPCService.java:15)
        at io.quarkus.devui.runtime.continuoustesting.ContinuousTestingJsonRPCService_ClientProxy.accept(Unknown Source)
        at io.quarkus.dev.testing.ContinuousTestingSharedStateManager.setLastState(ContinuousTestingSharedStateManager.java:36)
        at io.quarkus.deployment.dev.testing.ContinuousTestingSharedStateListener$1.runComplete(ContinuousTestingSharedStateListener.java:53)
        at io.quarkus.deployment.dev.testing.TestSupport.runInternal(TestSupport.java:475)
        at io.quarkus.deployment.dev.testing.TestSupport$3.run(TestSupport.java:381)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.RuntimeException: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class java.io.ByteArrayInputStream and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS) (through reference chain: io.quarkus.deployment.dev.testing.TestRunResults["results"]->java.util.Collections$UnmodifiableMap["AwesomeServiceTest"]->io.quarkus.deployment.dev.testing.TestClassResult["failing"]->java.util.ArrayList[0]->io.quarkus.deployment.dev.testing.TestResult["problems"]->java.util.Collections$UnmodifiableRandomAccessList[0]->org.jboss.resteasy.reactive.ClientWebApplicationException["cause"]->jakarta.ws.rs.WebApplicationException["response"]->org.jboss.resteasy.reactive.client.impl.ClientResponseImpl["entityStream"])
        at io.quarkus.devui.deployment.menu.ContinuousTestingProcessor.lambda$registerGetResultsMethod$5(ContinuousTestingProcessor.java:198)
        at io.quarkus.devui.runtime.continuoustesting.ContinuousTestingJsonRPCService.invokeAction(ContinuousTestingJsonRPCService.java:94)
        ... 9 more
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class java.io.ByteArrayInputStream and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS) (through reference chain: io.quarkus.deployment.dev.testing.TestRunResults["results"]->java.util.Collections$UnmodifiableMap["AwesomeServiceTest"]->io.quarkus.deployment.dev.testing.TestClassResult["failing"]->java.util.ArrayList[0]->io.quarkus.deployment.dev.testing.TestResult["problems"]->java.util.Collections$UnmodifiableRandomAccessList[0]->org.jboss.resteasy.reactive.ClientWebApplicationException["cause"]->jakarta.ws.rs.WebApplicationException["response"]->org.jboss.resteasy.reactive.client.impl.ClientResponseImpl["entityStream"])
        at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:77)
        at com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1306)
        at com.fasterxml.jackson.databind.DatabindContext.reportBadDefinition(DatabindContext.java:408)
        at com.fasterxml.jackson.databind.ser.impl.UnknownSerializer.failForEmpty(UnknownSerializer.java:53)
        at com.fasterxml.jackson.databind.ser.impl.UnknownSerializer.serialize(UnknownSerializer.java:30)
        at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:733)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774)
        at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
        at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:733)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774)
        at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
        at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:733)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774)
        at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:119)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:79)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:18)
        at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:733)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774)
        at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:119)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:79)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:18)
        at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:733)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774)
        at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
        at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
        at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
        at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
        at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
        at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:733)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774)
        at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
        at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4624)
        at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3852)
        at io.quarkus.devui.deployment.menu.ContinuousTestingProcessor.lambda$registerGetResultsMethod$5(ContinuousTestingProcessor.java:193)
        ... 11 more
```

### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.19045.3393]

### Output of `java -version`

openjdk version "17.0.5" 2022-10-18 OpenJDK Runtime Environment Temurin-17.0.5+8 (build 17.0.5+8) OpenJDK 64-Bit Server VM Temurin-17.0.5+8 (build 17.0.5+8, mixed mode, sharing)

### GraalVM version (if different from Java)

N/A

### Quarkus version or git rev

3.0.3.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.2 (c9616018c7a021c1c39be70fb2843d6f5f9b8a1c) Maven home: C:\Users\olson\dev\tools\apache-maven-3.9.2 Java version: 17.0.5, vendor: Eclipse Adoptium, runtime: C:\Program Files\Eclipse Adoptium\jdk-17.0.5.8-hotspot Default locale: de_DE, platform encoding: Cp1252 OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

### Additional information

N/A

--- a/extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/jsonrpc/ByteArrayInputStreamDeserializer.java
+++ b/extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/jsonrpc/ByteArrayInputStreamDeserializer.java
@@ -1,0 +1,27 @@
+package io.quarkus.devui.deployment.jsonrpc;
+
+import static io.quarkus.vertx.runtime.jackson.JsonUtil.BASE64_DECODER;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.exc.InvalidFormatException;
+
+public class ByteArrayInputStreamDeserializer extends JsonDeserializer<ByteArrayInputStream> {
+
+    @Override
+    public ByteArrayInputStream deserialize(JsonParser p, DeserializationContext ctxt)
+            throws IOException, JsonProcessingException {
+        String text = p.getText();
+        try {
+            byte[] decode = BASE64_DECODER.decode(text);
+            return new ByteArrayInputStream(decode);
+        } catch (IllegalArgumentException e) {
+            throw new InvalidFormatException(p, "Expected a base64 encoded byte array", text, ByteArrayInputStream.class);
+        }
+    }
+}

--- a/extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/jsonrpc/ByteArrayInputStreamSerializer.java
+++ b/extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/jsonrpc/ByteArrayInputStreamSerializer.java
@@ -1,0 +1,21 @@
+package io.quarkus.devui.deployment.jsonrpc;
+
+import static io.quarkus.vertx.runtime.jackson.JsonUtil.BASE64_ENCODER;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.SerializerProvider;
+
+import io.quarkus.deployment.util.IoUtil;
+
+public class ByteArrayInputStreamSerializer extends JsonSerializer<ByteArrayInputStream> {
+
+    @Override
+    public void serialize(ByteArrayInputStream value, JsonGenerator jgen, SerializerProvider provider) throws IOException {
+        byte[] readBytes = IoUtil.readBytes(value);
+        jgen.writeString(BASE64_ENCODER.encodeToString(readBytes));
+    }
+}

--- a/extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/jsonrpc/DevUIDatabindCodec.java
+++ b/extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/jsonrpc/DevUIDatabindCodec.java
@@ -1,5 +1,6 @@
 package io.quarkus.devui.deployment.jsonrpc;
 
+import java.io.ByteArrayInputStream;
 import java.io.Closeable;
 import java.io.IOException;
 import java.time.Instant;
@@ -137,6 +138,8 @@ public class DevUIDatabindCodec implements JsonMapper {
             module.addDeserializer(Instant.class, new InstantDeserializer());
             module.addSerializer(byte[].class, new ByteArraySerializer());
             module.addDeserializer(byte[].class, new ByteArrayDeserializer());
+            module.addSerializer(ByteArrayInputStream.class, new ByteArrayInputStreamSerializer());
+            module.addDeserializer(ByteArrayInputStream.class, new ByteArrayInputStreamDeserializer());
             mapper.registerModule(module);
 
             SimpleModule runtimeModule = new SimpleModule("vertx-module-runtime");

--- a/extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/menu/ContinuousTestingProcessor.java
+++ b/extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/menu/ContinuousTestingProcessor.java
@@ -1,12 +1,9 @@
 package io.quarkus.devui.deployment.menu;
 
 import java.io.IOException;
-import java.io.StringWriter;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Optional;
-
-import com.fasterxml.jackson.databind.ObjectMapper;
 
 import io.quarkus.arc.deployment.BeanContainerBuildItem;
 import io.quarkus.deployment.IsDevelopment;
@@ -233,7 +230,6 @@ public class ContinuousTestingProcessor {
 
     private void registerGetResultsMethod(LaunchModeBuildItem launchModeBuildItem) {
         DevConsoleManager.register(NAMESPACE + DASH + "getResults", ignored -> {
-            ObjectMapper objectMapper = new ObjectMapper(); // Remove in favior of build in.
             try {
                 Optional<TestSupport> ts = TestSupport.instance();
                 if (testsDisabled(launchModeBuildItem, ts)) {
@@ -246,10 +242,7 @@ public class ContinuousTestingProcessor {
                     return null;
                 }
 
-                try (StringWriter sw = new StringWriter()) {
-                    objectMapper.writeValue(sw, testRunResults);
-                    return sw.toString();
-                }
+                return testRunResults;
 
             } catch (Exception e) {
                 throw new RuntimeException(e);

--- a/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/testrunner/TestRunnerSmokeTestCase.java
+++ b/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/testrunner/TestRunnerSmokeTestCase.java
@@ -1,6 +1,6 @@
 package io.quarkus.vertx.http.testrunner;
 
-import java.util.List;
+import java.util.Iterator;
 import java.util.Map;
 import java.util.function.Function;
 import java.util.function.Supplier;
@@ -56,22 +56,21 @@ public class TestRunnerSmokeTestCase extends DevUIJsonRPCTest {
         Assertions.assertEquals(1L, ts.getTotalTestsPassed());
         Assertions.assertEquals(0L, ts.getTotalTestsSkipped());
 
-        JsonNode jsonRPCResultString = super.executeJsonRPCMethod("getResults");
-        Assertions.assertNotNull(jsonRPCResultString);
+        JsonNode jsonRPCResult = super.executeJsonRPCMethod("getResults");
+        Assertions.assertNotNull(jsonRPCResult);
+        Assertions.assertTrue(jsonRPCResult.has("results"));
 
-        @SuppressWarnings("unchecked")
-        Map<String, Map<String, Map>> jsonRPCResult = mapper.readValue(jsonRPCResultString.textValue(), Map.class);
-
-        Assertions.assertTrue(jsonRPCResult.containsKey("results"));
-
-        Map<String, Map> results = jsonRPCResult.get("results");
+        JsonNode results = jsonRPCResult.get("results");
         Assertions.assertNotNull(results);
-        for (Map.Entry<String, Map> result : results.entrySet()) {
-            @SuppressWarnings("unchecked")
-            Map<String, Object> testResult = result.getValue();
-            String className = (String) testResult.get("className");
-            List passing = (List) testResult.get("passing");
-            List failing = (List) testResult.get("failing");
+
+        Iterator<Map.Entry<String, JsonNode>> fields = results.fields();
+
+        while (fields.hasNext()) {
+            JsonNode testResult = fields.next().getValue();
+            String className = testResult.get("className").asText();
+
+            JsonNode passing = testResult.get("passing");
+            JsonNode failing = testResult.get("failing");
 
             if (className.equals(SimpleET.class.getName())) {
                 Assertions.assertEquals(1, failing.size(), className);

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/continuoustesting/ContinuousTestingJsonRPCService.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/continuoustesting/ContinuousTestingJsonRPCService.java
@@ -16,10 +16,10 @@ import io.smallrye.mutiny.operators.multi.processors.BroadcastProcessor;
 public class ContinuousTestingJsonRPCService implements Consumer<ContinuousTestingSharedStateManager.State> {
 
     private final BroadcastProcessor<String> stateBroadcaster = BroadcastProcessor.create();
-    private final BroadcastProcessor<String> resultBroadcaster = BroadcastProcessor.create();
+    private final BroadcastProcessor<Object> resultBroadcaster = BroadcastProcessor.create();
 
     private String lastKnownState = "";
-    private String lastKnownResults = "";
+    private Object lastKnownResults = "";
 
     @Override
     public void accept(ContinuousTestingSharedStateManager.State state) {
@@ -46,7 +46,7 @@ public class ContinuousTestingJsonRPCService implements Consumer<ContinuousTesti
         return stateBroadcaster;
     }
 
-    public Multi<String> streamTestResults() {
+    public Multi<Object> streamTestResults() {
         return resultBroadcaster;
     }
 
@@ -56,7 +56,7 @@ public class ContinuousTestingJsonRPCService implements Consumer<ContinuousTesti
     }
 
     @NonBlocking
-    public String lastKnownResults() {
+    public Object lastKnownResults() {
         return this.lastKnownResults;
     }
 
@@ -89,7 +89,7 @@ public class ContinuousTestingJsonRPCService implements Consumer<ContinuousTesti
         return invokeAction("toggleInstrumentation");
     }
 
-    public String getResults() {
+    public Object getResults() {
         return invokeAction("getResults");
     }
 

